catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/history/Historian.java

```diff
@@ -366,136 +242,169 @@ public void setCatalogProviders(List<CatalogProvider> catalogProviders) {
+        this.filterBuilder = filterBuilder;
     }
 
-    private CatalogProvider catalogProvider() {
-        return catalogProviders.stream()
-                .findFirst()
-                .orElse(null);
+    public void setSkipFlag(@Nullable Operation op) {
+        Optional.ofNullable(op)
+                .map(Operation::getProperties)
+                .ifPresent(p -> p.put(SKIP_VERSIONING, true));
+    }
+
+    private List<String> fromStorageRequests(Collection<ReadStorageRequest> requests) {
+        return requests.stream()
+                .map(StorageRequest::getId)
+                .collect(Collectors.toList());
     }
 
-    private void overwriteResourceUris(CreateStorageResponse createStorageResponse) {
-        createStorageResponse.getCreatedContentItems()
-                .forEach((ci) -> {
-                    try {
-                        ci.getMetacard()
-                                .setAttribute(new AttributeImpl(Metacard.RESOURCE_URI,
-                                        ci.getUri()));
-                        ci.getMetacard()
-                                .setAttribute(new AttributeImpl(Metacard.RESOURCE_SIZE,
-                                        ci.getSize()));
-                    } catch (IOException e) {
-                        LOGGER.debug("Could not get size", e);
-                    }
+    private Map<String, Metacard> query(Filter filter) throws UnsupportedQueryException {
+        SourceResponse response = catalogProvider().query(new QueryRequestImpl(new QueryImpl(filter,
+                1,
+                250,
+                null,
+                false,
+                TimeUnit.SECONDS.toMillis(10))));
 
-                });
+        return response.getResults()
+                .stream()
+                .map(Result::getMetacard)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toMap(Metacard::getId, Function.identity()));
     }
 
-    private List<ContentItem> versionContentItems(List<ContentItem> contentItems,
-            Map<String, Path> tmpContentPaths, MetacardVersion.Action action) {
-        return contentItems.stream()
+    private Filter forIds(List<String> ids) {
+        List<Filter> idFilters = ids.stream()
+                .map(id -> filterBuilder.attribute(Metacard.ID)
+                        .is()
+                        .equalTo()
+                        .text(id))
+                .collect(Collectors.toList());
+
+        return filterBuilder.anyOf(idFilters);
+    }
+
+    private Map<String, List<ContentItem>> getOldContent(Collection<ReadStorageRequest> ids) {
+        return ids.stream()
+                .map(this::getStorageItem)
+                .filter(Objects::nonNull)
+                .map(ReadStorageResponse::getContentItem)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toMap(ContentItem::getId,
+                        ImmutableList::of,
```

Comment:
Are these going to be large collections of ids? You're generating and copying a potentially very large number of ImmutableLists this way. Feels like a map/reduce operation might work better, or even just using a mutable collection here and then wrapping the final result in a unmodifiable. This is a lot of object creation/gc.